atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -1769,18 +1783,42 @@ protected boolean shouldDeleteAndRollback() {
 
     @Override
     public final void delete(TableReference tableRef, Set<Cell> cells) {
+        deleteInternal(tableRef, cells, ImmutableMap.of());
+    }
+
+    @Override
+    public void deleteWithMetadata(TableReference tableRef, Map<Cell, ChangeMetadata> cellsWithMetadata) {
+        deleteInternal(tableRef, cellsWithMetadata.keySet(), cellsWithMetadata);
+    }
+
+    private void deleteInternal(TableReference tableRef, Set<Cell> cells, Map<Cell, ChangeMetadata> metadata) {
         getCache().delete(tableRef, cells);
-        putInternal(tableRef, Cells.constantValueMap(cells, PtBytes.EMPTY_BYTE_ARRAY));
+        putInternal(tableRef, Cells.constantValueMap(cells, PtBytes.EMPTY_BYTE_ARRAY), metadata);
     }
 
     @Override
     public void put(TableReference tableRef, Map<Cell, byte[]> values) {
+        putWithMetadataInternal(tableRef, values, ImmutableMap.of());
+    }
+
+    @Override
+    public void putWithMetadata(
+            TableReference tableRef, Map<Cell, ChangeMetadataAnnotatedValue> metadataAnnotatedValues) {
+        Map<Cell, byte[]> valuesOnly =
+                Maps.transformValues(metadataAnnotatedValues, ChangeMetadataAnnotatedValue::value);
+        Map<Cell, ChangeMetadata> metadataOnly =
+                Maps.transformValues(metadataAnnotatedValues, ChangeMetadataAnnotatedValue::metadata);
+        putWithMetadataInternal(tableRef, valuesOnly, metadataOnly);
+    }
+
+    private void putWithMetadataInternal(
+            TableReference tableRef, Map<Cell, byte[]> values, Map<Cell, ChangeMetadata> metadata) {
         ensureNoEmptyValues(values);
         getCache().write(tableRef, values);
-        putInternal(tableRef, values);
+        putInternal(tableRef, values, metadata);
     }
 
-    public void putInternal(TableReference tableRef, Map<Cell, byte[]> values) {
+    public void putInternal(TableReference tableRef, Map<Cell, byte[]> values, Map<Cell, ChangeMetadata> metadata) {
```

Comment:
Now we basically have two `putInternals` (one that does validation and deals with the cache and one that writes to our local buffer). I called the first one `putWithMetadataInternal`, but I don't think that is the best name. I would like to have a dedicated method to handle the validation and caches so we are more resilient against future changes to `put` or `putWithMetadata`.